src/compiler/scala/tools/nsc/symtab/classfile/JavaSignatureWalker.scala

```diff
@@ -0,0 +1,108 @@
+package scala.tools.nsc.symtab.classfile
+
+import scala.reflect.internal.ClassfileConstants._
+
+abstract class JavaSignatureWalker {
+  def visitName(internalName: CharSequence): Unit
+  def raiseError(msg: String): Unit
+
+  def walk(sig: CharSequence): Unit = {
+    def raiseError(msg: String): Unit = {
+      this.raiseError(s"$msg (while parsing $sig)")
+    }
+    var index = 0
+    val end = sig.length
+    def accept(ch: Char) {
+      assert(current() == ch, (current(), ch))
+      index += 1
+    }
+    def skipSubName(isDelimiter: Char => Boolean): Unit = {
+      val start = index
+      while (!isDelimiter(current())) { index += 1 }
+    }
+    def appendSubName(builder: java.lang.StringBuilder, isDelimiter: Char => Boolean): CharSequence = {
+      val start = index
+      skipSubName(isDelimiter)
+      builder.append(sig, start, index)
+    }
+    def current(): Char = {
+      if (index < sig.length())
```

Comment:
`sig.length` / `end`. I would prefer if `current` throws when the index is out of bounds (fail fast).